### rust in bioinformatics

A collection of genomics software tools written in Rust



### index section

##### bam

- [best](https://github.com/google/best) : Bam Error Stats Tool (best): analysis of error types in aligned reads
- [modkit](https://github.com/nanoporetech/modkit) : A bioinformatics tool for working with modified bases
- [mapAD](https://github.com/mpieva/mapAD) : An aDNA aware short-read mapper
- [perbase](https://github.com/sstadick/perbase) : Per-base per-nucleotide depth analysis
- [rustybam](https://github.com/mrvollger/rustybam) : bioinformatics toolkit in rust

##### csv

- [csview](https://github.com/wfxr/csview) : üì† Pretty and fast csv viewer for cli with cjk/emoji support
- [csvlens](https://github.com/YS-L/csvlens) : csvlens is a command line CSV file viewer. It is like less but made for CSV.
- [madato](https://github.com/inosion/madato) : Markdown Cmd Line, Rust and JS library for Excel to Markdown Tables
- [tabiew](https://github.com/shshemi/tabiew) : A lightweight TUI app to view and query CSV files
- [tv](https://github.com/alexhallam/tv) : üì∫(tv) Tidy Viewer is a cross-platform CLI csv pretty printer that uses column styling to maximize viewer enjoyment.
- [xsv](https://github.com/BurntSushi/xsv) : A fast CSV command line toolkit written in Rust. ¬†
- [xtab](https://github.com/sharkLoc/xtab) : CSV command line utilities

##### dna

- [fakit](https://github.com/sharkLoc/fakit) : fakit: a simple program for fasta file manipulation
- [fq](https://github.com/stjude-rust-labs/fq) : Command line utility for manipulating Illumina-generated FASTQ files.
- [ngs](https://github.com/stjude-rust-labs/ngs) : Command line utility for working with next-generation sequencing files.
- [poasta](https://github.com/broadinstitute/poasta) : Fast and exact gap-affine partial order alignment
- [rust-bio-tools](https://github.com/rust-bio/rust-bio-tools) : A set of command line utilities based on Rust-Bio.
- [skc](https://github.com/mbhall88/skc) : Shared k-mer content between two genomes

##### fastq

- [fasten](https://github.com/lskatz/fasten) : üë∑ Fasten toolkit, for streaming operations on fastq files
- [faster](https://github.com/angelovangel/faster) :  A (very) fast program for getting statistics about a fastq file, the way I need them, written in Rust
- [fqgrep](https://github.com/fulcrumgenomics/fqgrep) : Grep for FASTQ files
- [fqkit](https://github.com/sharkLoc/fqkit) : ü¶Ä Fqkit: A simple and cross-platform program for fastq file manipulation ¬†
- [fqtk](https://github.com/fulcrumgenomics/fqtk) : Fast FASTQ sample demultiplexing in Rust.
- [rasusa](https://github.com/mbhall88/rasusa) : Randomly subsample sequencing reads

#### format
- [bigtools](https://github.com/jackh726/bigtools) : A high-performance BigWig and BigBed library in Rust
- [d4tools](https://github.com/38/d4-format) : The D4 Quantitative Data Format
- [thirdkind](https://github.com/simonpenel/thirdkind) : Drawing reconciled phylogenetic trees allowing 1, 2 or 3 reconcillation levels


##### gff3

- [gffkit](https://github.com/sharkloc/gffkit) : a simple program for gff3 file manipulation

##### longreads

- [chopper](https://github.com/wdecoster/chopper) : Rust implementation of [NanoFilt](https://github.com/wdecoster/nanofilt)+[NanoLyse](https://github.com/wdecoster/nanolyse), both originally written in Python. This tool, intended for long read sequencing such as PacBio or ONT, filters and trims a fastq file.
- [longshot](https://github.com/pjedge/longshot) : diploid SNV caller for error-prone reads
- [nextpolish2](https://github.com/Nextomics/NextPolish2) : Repeat-aware polishing genomes assembled using HiFi long reads

##### metagenomics

- [coverm](https://github.com/wwood/CoverM) : Read coverage calculator for metagenomics
- [skani](https://github.com/bluenote-1577/skani) : Fast, robust ANI and aligned fraction for (metagenomic) genomes and contigs.
- [sylph](https://github.com/bluenote-1577/sylph) : ultrafast genome querying and taxonomic profiling for metagenomic samples by abundance-corrected minhash.
- [vircov](https://github.com/esteinig/vircov) : Viral genome coverage evaluation for metagenomic diagnostics ü©∏

##### pangenomics

- [impg](https://github.com/pangenome/impg) : implicit pangenome graph

##### phylogenomics

- [segul](https://github.com/hhandika/segul) : An ultrafast and memory efficient tool for phylogenomics

##### proteomics

- [sage](https://github.com/lazear/sage) : Proteomics search & quantification so fast that it feels like magic

##### rna

- [rnapkin](https://github.com/ukmrs/rnapkin) : drawing RNA secondary structure with style; instantly
- [squab](https://github.com/zaeleus/squab) : Alignment-based gene expression quantification

##### singlecell

- [alevin-fry](https://github.com/COMBINE-lab/alevin-fry) : üêü üî¨ü¶Ä alevin-fry is an efficient and flexible tool for processing single-cell sequencing data, currently focused on single-cell transcriptomics and feature barcoding.
- [cellranger](https://github.com/10XGenomics/cellranger) : 10x Genomics Single Cell Analysis

##### slurm

- [ssubmit](https://github.com/mbhall88/ssubmit) : Submit slurm sbatch jobs without the need to create a script

##### vcf

- [echtvar](https://github.com/brentp/echtvar) : using all the bits for echt rapid variant annotation and filtering
- [vcf2parquet](https://github.com/natir/vcf2parquet) : Convert vcf in parquet
